File: src/lib/debrid/registry.ts
import { useMemo } from "react";
import { useDebridSecrets } from "@/lib/secret-store";
import { createAllDebrid } from "./alldebrid";
import { createDebridLink } from "./debridlink";
import { createPremiumize } from "./premiumize";
import { createRealDebrid } from "./realdebrid";
import { createTorbox } from "./torbox";
import type { DebridSlug, DebridStore } from "./types";

export type DebridKeys = {
  rdKey: string;
  tbKey: string;
  adKey: string;
  pmKey: string;
  dlKey: string;
};

export function buildDebridClients(keys: DebridKeys): DebridStore[] {
  const clients: DebridStore[] = [];
  if (keys.rdKey) clients.push(createRealDebrid(keys.rdKey));
  if (keys.tbKey) clients.push(createTorbox(keys.tbKey));
  if (keys.adKey) clients.push(createAllDebrid(keys.adKey));
  if (keys.pmKey) clients.push(createPremiumize(keys.pmKey));
  if (keys.dlKey) clients.push(createDebridLink(keys.dlKey));
  return clients;
}

export function useDebridClients(): DebridStore[] {
  const keys = useDebridSecrets();
  return useMemo(
    () => buildDebridClients(keys),
    [keys.rdKey, keys.tbKey, keys.adKey, keys.pmKey, keys.dlKey],
  );
}

export function useDebridSlugs(): DebridSlug[] {
  const clients = useDebridClients();
  return useMemo(() => clients.map((c) => c.slug), [clients]);
}

File: clients/lib/domain/debrid/registry.dart
import '../streams/parser/stream_enums.dart';
import 'alldebrid.dart';
import 'debrid_http.dart';
import 'debrid_link.dart';
import 'debrid_types.dart';
import 'premiumize.dart';
import 'real_debrid.dart';
import 'torbox.dart';

/// The five debrid API keys from settings (each defaults empty). Ported from
/// `src/lib/debrid/registry.ts`.
class DebridKeys {
  const DebridKeys({
    this.rdKey = '',
    this.tbKey = '',
    this.adKey = '',
    this.pmKey = '',
    this.dlKey = '',
  });

  final String rdKey;
  final String tbKey;
  final String adKey;
  final String pmKey;
  final String dlKey;
}

/// Builds a debrid client for each non-empty key, in the fixed order rd, tb, ad,
/// pm, dl — the default resolution order before per-stream cache re-sorting.
List<DebridStore> buildDebridClients(DebridKeys keys, DebridHttp http) {
  final clients = <DebridStore>[];
  if (keys.rdKey.isNotEmpty) clients.add(RealDebrid(keys.rdKey, http));
  if (keys.tbKey.isNotEmpty) clients.add(Torbox(keys.tbKey, http));
  if (keys.adKey.isNotEmpty) clients.add(AllDebrid(keys.adKey, http));
  if (keys.pmKey.isNotEmpty) clients.add(Premiumize(keys.pmKey, http));
  if (keys.dlKey.isNotEmpty) clients.add(DebridLink(keys.dlKey, http));
  return clients;
}

/// The slugs of the configured debrid clients, in build order.
List<DebridSlug> debridSlugs(List<DebridStore> clients) =>
    clients.map((c) => c.slug).toList();
